### PR TITLE
style & refactor : improvement of Group UI/UX

### PR DESCRIPTION
**Pull request by bivanalhar**
Several improvements made for the Group page are as follows:

1. Main Group Page
- Include the indicator icon if the user is a Phantom user
- Include the Edit functionality button in each of the groups in the main page, that will direct user towards the editing of that particular group.
- Changing of the icon for Member and Manager in the main page, to make it easier to differentiate between two of them
- Include the checkbox on the bottom to each group, which has the functionality to hide all the phantom users inside the group

Before
Screenshot 2023-01-20 at 2 06 30 PM

After
Screenshot 2023-01-20 at 2 09 46 PM

Hide Phantom checklist functionality
Screenshot 2023-01-20 at 2 12 05 PM

2. Edit Group Page
- Include the checkbox to hide all Phantom users, both in "Users that can be added" and "Users in group"
- Disable the changing of the group role for students by changing the dropdown into the Member icon
- Indicate if user is belonging to another group, as well as to which groups they are belonging to

Before
Screenshot 2023-01-20 at 2 13 44 PM

After
Screenshot 2023-01-20 at 2 49 18 PM

Hide Phantom checkbox functionality
Screenshot 2023-01-20 at 2 51 00 PM

3. Miscellanous
- Remove the warnings in the propTypes when some fields are required but value is undefined

### CODE REVIEW

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx [hunk 1]
```diff
@@ -107,49 +163,41 @@ const GroupUserManagerListItem = ({
 
       <ListItemText primaryTypographyProps={{ style: styles.listItemTextSize }}>
         {user.name}
+        {user.isPhantom ? <GhostIcon /> : ''}
+        {otherGroups ? ` (also a member of${otherGroups})` : ''}
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:167]: extract `also a member of` as intl

File: client/app/bundles/course/group/pages/GroupShow/index.jsx [hunk 2]
```diff
@@ -93,7 +101,15 @@ const Category = ({
             }
           />
           {groups.map((group) => (
-            <GroupTableCard key={group.id} group={group} />
+            <GroupTableCard
+              key={group.id}
+              canManageCategory={canManageCategory}
+              group={group}
+              onManageGroups={() => {
+                dispatch({type: actionTypes.MANAGE_GROUPS_START})
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/index.jsx:109]: You can put this together inside handleGroupSelect so that the dispatches are grouped together. 

Also, any reason it's called `onManageGroups`? I understand that this function is to go edit a single group, so perhaps `onManageGroup`?

Finally, you should scroll to the top of <GroupUserManager /> component when the manage group button is clicked. Try having many groups and you click the last manage group. After clicking and redirecting to the page, you are now at the bottom at the page.

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx [hunk 3]
```diff
@@ -98,12 +158,24 @@ const GroupTableCard = ({ group }) => {
           <FormattedMessage {...translations.noMembers} />
         </div>
       ) : null}
+      <FormControlLabel
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:161]: This should be placed before the table. Let's say we have a very long group table, then we need to scroll to the bottom of the table to find this checkbox

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx [hunk 4]
```diff
@@ -35,11 +43,31 @@ const translations = defineMessages({
     defaultMessage:
       'This group has no members! Manage groups to assign members now!',
   },
+  manageOneGroup: {
+    id: 'course.group.GroupShow.CategoryCard.manageOne',
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:47]: Be careful with the key. After you paste it, dont forget to change the id.. See the patterns. We cant have duplicate ids across our codebase for intl.

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx [hunk 5]
```diff
@@ -85,9 +131,23 @@ const GroupTableCard = ({ group }) => {
...
-              <TableCell style={styles.rowHeight}>{m.name}</TableCell>
               <TableCell style={styles.rowHeight}>
-                {roles[m.groupRole]}
+                <div className="flex grow">
+                  {m.name}
+                  {m.isPhantom ? <GhostIcon /> : ''}
+                </div>
+              </TableCell>
+              <TableCell style={styles.rowHeight}>
+                <Chip
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:141]: I think this is reused in `GroupManagerList` right? You can abstract this into a reusable component

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx [hunk 6]
```diff
@@ -85,9 +131,23 @@ const GroupTableCard = ({ group }) => {
           {members.map((m, index) => (
             <TableRow key={m.id} style={styles.rowHeight}>
               <TableCell style={styles.rowHeight}>{index + 1}</TableCell>
-              <TableCell style={styles.rowHeight}>{m.name}</TableCell>
               <TableCell style={styles.rowHeight}>
-                {roles[m.groupRole]}
+                <div className="flex grow">
+                  {m.name}
+                  {m.isPhantom ? <GhostIcon /> : ''}
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:137]: For brevity, `{m.isPhantom && <GhostIcon /> }`

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx [hunk 7]
```diff
@@ -81,12 +102,47 @@ const styles = {
   },
 };
 
+const GroupUserManagerListItemChoice = ({ user, onChangeDropdown }) =>
+  user.role !== 'student' ? (
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:106]: As much as possible, we try to avoid negative conditionals. So u can swap it to `user.role == 'student'`

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx [hunk 8]
```diff
@@ -107,49 +163,41 @@ const GroupUserManagerListItem = ({
...
   staff = [],
+  memberOtherGroups = {},
   onCheck,
   colourMap,
   showDropdown = false,
   onChangeDropdown,
   isChecked = false,
 }) => {
-  const renderUsersListItems = (users, title) => (
+  const renderUsersListItems = (users, members, title) => (
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:200]: Is this supposed to be `memberOtherGroups`?

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 9]
```diff
@@ -118,22 +130,65 @@ const getAvailableUsers = (
   groups,
   group,
   hideInGroup,
+  hidePhantomStudent,
   availableSearch,
 ) => {
+  let groupMemberIds = new Set()
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:136]: I think a cleaner way to do this is to start with the full list of the group member ids and sequentially remove unfiltered ids. You can try it first and I will improve later.

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 10]
```diff
@@ -118,22 +130,65 @@ const getAvailableUsers = (
...
+  )
+};
+
+const getAvailableUserInOtherGroups = (
+  courseUsers,
+  groups,
+  group,
+) => {
+  const otherGroups = groups.filter((x) => x !== group)
+  const mapStudentToGroups = {}
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:178]: It's better to store the value of the object as an array containing the group names. When needed later, we can just use `join`.

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 11]
```diff
@@ -118,22 +130,65 @@ const getAvailableUsers = (
...
+    groupMembers = new Set(members)
+  }
+
+  return filterByName(
+    selectedSearch,
+    [...groupMembers],
+  )
+};
+
+const getAvailableUserInOtherGroups = (
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:172]: Actually I suggest this to be done on the backend side.
Maybe `json.otherGroupIds` to be added to `json.members` in `coursemology2/app/views/course/group/_group.json.jbuilder`. This way, we can reduce the complexity in the frontend and let the backend do the heavy lifting. Also, since there's no other interaction that will cause changes to the frontend view dynamically, it makes sense to do it in the backend. (And most importantly for you to work on some backend code)

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Good attempt! @bivanalhar Please also change to tailwind for the styling.

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx [hunk 12]
```diff
@@ -36,6 +40,23 @@ const translations = defineMessages({
   },
 });
 
+const groupRoleTranslation = {
+  normal: 'Member',
+  manager: 'Manager',
+  unknown: 'Unknown',
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:46]: thanks a lot for notifying this. actually at first when I was doing this, the code has some bug (somehow the role does not fall into member and manager) and thus that unknown stuff is merely for debugging. But now it's been finished, the role is no longer needed but I forgot to remove this.. will remove it soon

File: client/app/theme/palette.js [hunk 13]
```diff
@@ -78,6 +79,13 @@ const palette = {
     [workflowStates.Graded]: colors.blue[100],
     [workflowStates.Published]: colors.green[100],
   },
+
+  groupRole: {
+    [groupRole.Normal]: colors.green[100],
+    [groupRole.Manager]: colors.red[100],
+    [groupRole.Unknown]: colors.grey[100],
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/theme/palette.js:86]: same as above (will be removed soon)

File: client/app/bundles/course/group/pages/GroupShow/CategoryCard.jsx [hunk 14]
```diff
@@ -133,7 +133,7 @@ const CategoryCard = ({
         onClick: handleEdit,
       });
     }
-    if (canManageGroups) {
+    if (canManageCategory) {
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/CategoryCard.jsx:136]: previously, the permission for canManageGroups also allow students (who are the manager, previously made possible) to see the page containing members of the groups. But now, since students are no longer able to become a manager, and also no longer able to see the page, then basically this permission setting is no longer required.

I will remove all the canManageGroups related stuffs from here

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 15]
```diff
@@ -118,22 +130,65 @@ const getAvailableUsers = (
   groups,
   group,
   hideInGroup,
+  hidePhantomStudent,
   availableSearch,
 ) => {
+  let groupMemberIds = new Set()
+  // const altGroup = new Set(groups.flatMap((g) => g.members).filter((m) => !m.isPhantom));
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:137]: if I didn't set this, then the groupMemberIds will only be defined within the if..else.. statement. But, I have found a way by using ternary to make the variable assignment done only once

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 16]
```diff
@@ -414,6 +496,17 @@ const GroupUserManager = ({
           label={<FormattedMessage {...translations.hideStudents} />}
           style={styles.checkbox}
         />
+        <br />
+        <FormControlLabel
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:500]: roger that.. done and will be committed in the next commit

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 17]
```diff
@@ -281,6 +356,11 @@ const GroupUserManager = ({
   const onChangeRole = useCallback(
     (value, user) => {
       if (user.groupRole === value) return undefined;
+      if (user.role === "student" && value === "manager") {
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:359]: ah yes you're right. I have removed this one.
As for the practice, yes I agree with you that better backend handle this process. However, I am still not that familiar with how to practice that (I'm still learning on how to do it properly). Surely after I catch up with the knowledge on how to do that, I will do the most efficient way to implement this kind of thing

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/NonGroupUserManagerList.jsx [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+  staff: PropTypes.arrayOf(nonMemberShape),
+  memberOtherGroups: PropTypes.arrayOf(nonMemberShape),
+  onCheck: PropTypes.func.isRequired,
+  colourMap: PropTypes.object.isRequired,
+  showDropdown: PropTypes.bool,
+  onChangeDropdown: PropTypes.func,
+  isChecked: PropTypes.bool,
+};
+
+export default NonGroupUserManagerList;
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/NonGroupUserManagerList.jsx:223]: for this one, I would like to seek your advice regarding this. To be frank, I have tried to combine these 2 files before, for the exact same reason as you mentioned (too many same functionalities between these 2 files). However, the difference between those 2 lies in a quite fundamental way (the propTypes have one difference and so whenever I want to implement, there will be errors due to one variable defined but other undefined, sth like that)

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx (hunk 1 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:167]: this one, I have tried but somehow it produces weird error (it detects the id of the translation instead of its defaultMessages). I will look into it later

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx (hunk 3 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:161]: okay

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx (hunk 6 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:137]: okay

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx (hunk 7 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:106]: okay

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx (hunk 8 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:200]: this is true.. I just deleted the members variable since it's not used anywhere inside the function itself

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx (hunk 9 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:136]: I am using the ternary operation instead. Including it in the commit and maybe later we can see which one actually is better?

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx (hunk 10 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:178]: actually that mapStudentToGroups does not only contain info of groups attended by one student. It has the informations of each student joining which groups. i don't think array is the good data structure to be used here

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx (hunk 5 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:141]: need more advise regarding this. tried to abstract this into reusable component, but seems like the value `user` is not passed over to the function

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/NonGroupUserManagerList.jsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/NonGroupUserManagerList.jsx]: okay it's done already

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx]: guess can as well mark this as resolved, since the checking is no longer there

File: client/app/bundles/course/group/pages/GroupShow/index.jsx (hunk 2 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/index.jsx:109]: `Finally, you should scroll to the top of component when the manage group button is clicked. Try having many groups and you click the last manage group. After clicking and redirecting to the page, you are now at the bottom at the page.`

can you please elaborate a bit more on this @ekowidianto ?? I tried to reproduce the mentioned things here and seems like it's not necessarily going down to the bottom

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx (hunk 5 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx]: done

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx (hunk 1 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx]: done

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx (hunk 10 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx]: resolved

File: client/app/bundles/course/group/pages/GroupShow/index.jsx (hunk 2 above)
Comment (bivanalhar) [reply client/app/bundles/course/group/pages/GroupShow/index.jsx]: done

File: client/app/bundles/course/group/constants.js [hunk 19]
```diff
@@ -9,6 +9,12 @@ export const dialogTypes = mirrorCreator([
   'UPDATE_GROUP',
 ]);
 
+export const groupRole = {
+  Normal: 'normal',
+  Manager: 'manager',
+  Unknown: 'unknown',
```
Comment (ekowidianto) [inline client/app/bundles/course/group/constants.js:15]: Dont forget to remove

File: client/app/bundles/course/group/components/GroupRoleChip.jsx [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+  manager: 'Manager',
+};
+
+const translateStatus = (oldStatus) => {
+  switch (oldStatus) {
+    case 'normal':
+      return groupRoleTranslation.normal;
+    case 'manager':
+      return groupRoleTranslation.manager;
+    default:
```
Comment (ekowidianto) [inline client/app/bundles/course/group/components/GroupRoleChip.jsx:17]: Dont forget to remove this also.. Instead of using switch, use dict

File: client/app/bundles/course/group/pages/GroupShow/CategoryCard.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/bundles/course/group/pages/GroupShow/CategoryCard.jsx:135]: Have you double checked the logic of `canManageCategory` and `canManageGroups` which are derived from the backend side?

**Comment by ekowidianto** [comment]
@bivanalhar please change to tailwind for the styling of the files that you change in this PR.

File: client/app/bundles/course/group/pages/GroupShow/index.jsx [hunk 21]
```diff
@@ -88,12 +96,20 @@ const Category = ({
...
+            }}
           />
           {groups.map((group) => (
-            <GroupTableCard key={group.id} group={group} />
+            <GroupTableCard
+              key={group.id}
+              canManageCategory={canManageCategory}
+              group={group}
+              onManageGroup={() => {
+                dispatch({ type: actionTypes.MANAGE_GROUPS_START });
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/index.jsx:109]: Any reason why this is not placed inside handleGroupSelect?

File: client/app/bundles/course/group/pages/GroupShow/index.jsx [hunk 22]
```diff
@@ -88,12 +96,20 @@ const Category = ({
             canManageGroups={canManageGroups}
             category={groupCategory}
             numGroups={groups.length}
-            onManageGroups={() =>
-              dispatch({ type: actionTypes.MANAGE_GROUPS_START })
-            }
+            onManageGroups={() => {
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/index.jsx:99]: Revert this change

File: client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx [hunk 23]
```diff
@@ -35,13 +42,16 @@ const translations = defineMessages({
...
+    defaultMessage: 'Hide all phantom students',
+  },
 });
 
-const roles = {
-  normal: 'Member',
-  manager: 'Manager',
-};
-
 const styles = {
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupTableCard.jsx:55]: Change to tailwind

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx [hunk 24]
```diff
@@ -66,6 +67,14 @@ const translations = defineMessages({
     defaultMessage:
       'Hide students who are already in a group under this category',
   },
+  hidePhantomStudents: {
+    id: 'course.group.GroupShow.GroupManager.GroupUserManager.hidePhantomStudents',
+    defaultMessage: 'Hide all Phantom Students',
+  },
+  cannotUpliftStudent: {
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx:74]: Is this still used?

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx [hunk 25]
```diff
@@ -81,13 +90,49 @@ const styles = {
   },
 };
 
+const GroupUserManagerListItemChoice = ({
+  isOutsideGroup,
+  user,
+  onChangeDropdown,
+}) => {
+  if (isOutsideGroup || !user.groupRole) return undefined;
+  return user.role === 'student' ? (
```
Comment (ekowidianto) [inline client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManagerList.jsx:99]: Do we need this role check? They are not supposed to be able to enter this page no?

File: client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx (hunk 24 above)
Comment (ekowidianto) [reply client/app/bundles/course/group/pages/GroupShow/GroupManager/GroupUserManager.jsx]: I have removed it for you

File: client/app/bundles/course/group/components/GroupRoleChip.jsx (hunk 20 above)
Comment (ekowidianto) [reply client/app/bundles/course/group/components/GroupRoleChip.jsx]: I have refactored this. Refer to the latest changes.

File: client/app/bundles/course/group/constants.js (hunk 19 above)
Comment (ekowidianto) [reply client/app/bundles/course/group/constants.js]: Removed for you also~

**Comment by ekowidianto** [comment]
@bivanalhar You can have a look at my commits and let me know if there's any question~

**Comment by bivanalhar** [comment]
looks good to me!!! Thanks a lot @ekowidianto
